text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+Enum::ThirdVariant { x: 42 };
+```
+
+Relative to `#[derive(new)]`, the main benefits are:
+
++ No wrapping code in strings, as noted in previous sections.
++ The defaults used can be mixed and matches; it works to request all defaults
+  or just some of them.
+
+The constructor functions `new_first_variant(..)` are not provided for you.
+However, it should be possible to tweak `#[derive(new)]` to interact with
+this RFC so that constructor functions are regained if so desired.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+1. What is the right interaction wrt. `#[non_exhaustive]`?
+
+   In particular, if given the following definition:
+
+   ```rust
+   #[non_exhaustive]
+   pub struct Config {
+       pub height: u32,
+       pub width: u32,
+   }
+   ```
+
+   it is possible to construct a `Config` like so:
+
+   ```rust
+   let config = Config { width: 640, height: 480, .. };
+   ```
+
+   then adding a field to `Config` can only happen iff that field
+   is provided a default value.
+
+   This arrangement, while diminishing the usefulness of `#[non_exhaustive]`,
+   makes the ruleset of the language simpler, more consistent, and also
+   simplifies type checking as `#[non_exhaustive]` is entirely ignored
+   when checking `Foo { fields, .. }` expressions.
+
+   As an alternative, users who desire the semantics described above can
+   omit `#[non_exhaustive]` from their type and instead add a private
+   defaulted field that has a ZST.
+
+   Resolving this question may be deferred to stabilization of either
+   `#[non_exhaustive]` or field defaults, whichever comes last.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Tuple structs and tuple variants
+
+Although it could, this proposal does not offer a way to specify default values
+for tuple struct / variant fields. For example, you may not write:
+
+```rust
+#[derive(Default)]
+struct Alpha(u8 = 42, bool = true);
+
+#[derive(Default)]
+enum Ingredient {
+    Tomato(TasteQuality, Color = Color::Red),
+    Lettuce,
+}
+```
+
+While well-defined semantics could be given for these positional fields,
+there are some tricky design choices; in particular:
+
++ It's unclear whether the following should be permitted:
+
+  ```rust
+  #[derive(Default)]
+  struct Beta(&'static str = "hello", bool);
+  ```
+
+  In particular, the fields with defaults are not at the end of the struct.
+  A restriction could imposed to enforce that. However, it would also be
+  useful to admit the above definition of `Beta` so that `#[derive(Default)]`
+  can make use of `"hello"`.
+
++ The syntax `Alpha(..)` as an expression already has a meaning.
+  Namely, it is sugar for `Alpha(RangeFull)`. Thus unfortunately,
+  this syntax cannot be used to mean `Alpha(42, true)`.
+  In some future edition, the syntax `Alpha(...)` (three dots)
+  could be used for filling in defaults. This would ostensibly entail
+  adding the pattern syntax `Alpha(...)` as well.
```

Comment:
I believe it is still possible to reuse the `..` syntax, though the complexity of making it work in the compiler might be undesirable. I'm also not a fan of using `...` in tuple structs and `..` in regular structs. I think they should be consistent.

But we do have `...` available today in current editions.

There is another possibility not mentioned here, which is using `_` for defaulted fields (though it could equally apply to regular structs).

Reply:
```suggestion
+ The syntax `Alpha(..)` as an expression already has a meaning.
  Namely, it is sugar for `Alpha(RangeFull)`. Thus unfortunately,
  this syntax cannot be used to mean `Alpha(42, true)`.
  In newer editions, the syntax `Alpha(...)` (three dots)
  can be used for filling in defaults. This would ostensibly entail
  adding the pattern syntax `Alpha(...)` as well.
```